arangod/VocBase/Methods/AqlUserFunctions.cpp

```diff
@@ -0,0 +1,445 @@
+  trx.addHint(transaction::Hints::Hint::SINGLE_OPERATION);
+
+  res = trx.begin();
+  if (!res.ok()) {
+    return res;
+  }
+
+  std::unique_ptr<ManagedDocumentResult> mmdr;
+  transaction::BuilderLeaser searchBuilder(&trx);
+  searchBuilder->add(VPackValue(_key));
+  arangodb::velocypack::Builder existDocument;
+  try {
+    res = trx.documentFastPath(collectionName, mmdr.get(), searchBuilder->slice(), existDocument, false);
+  } catch (arangodb::basics::Exception const& ex) {
+    res.reset(ex.code());
+  }
+
+  {
+    arangodb::OperationResult result;
+    replacedExisting = res.ok();
+    if (replacedExisting){
+      result = trx.replace(collectionName, oneFunctionDocument.slice(), opOptions);
+    } else {
+      result = trx.insert(collectionName, oneFunctionDocument.slice(), opOptions);
+    }
+    // Will commit if no error occured.
+    // or abort if an error occured.
+    // result stays valid!
+    res = trx.finish(result.result);
+  }
+
+  if (res.ok()) {
+    reloadAqlUserFunctions();
+  }
+
+  return res;
+}
+
+Result arangodb::toArrayUserFunctions(TRI_vocbase_t* vocbase,
+                                      std::string const& functionFilterPrefix,
+                                      velocypack::Builder& result) {
+  auto binds = std::make_shared<VPackBuilder>();
+  binds->openObject();
+  if (! functionFilterPrefix.empty()) {
+    std::string uc(functionFilterPrefix);
+    basics::StringUtils::toupperInPlace(&uc);
+    if ((uc.length() < 2) ||
+        (uc[uc.length() - 1 ] != ':') ||
+        (uc[uc.length() - 2 ] != ':')) {
+      uc += "::";
+    }
+    binds->add("fnLength", VPackValue(uc.length()));
+    binds->add("ucName", VPackValue(uc));
+  }
+  binds->add("@col", VPackValue(collectionName));
+  binds->close();
+
+  std::string const aql(
+      "FOR function IN @@col "
+      "FILTER LEFT(function._key, @fnLength) == @ucName "
+      "RETURN function"
+  );
+
+  arangodb::aql::Query query(false, vocbase, arangodb::aql::QueryString(aql),
+                             binds, nullptr, arangodb::aql::PART_MAIN);
+
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  auto queryResult = query.execute(queryRegistry);
+
+  if (queryResult.code != TRI_ERROR_NO_ERROR) {
+    if (queryResult.code == TRI_ERROR_REQUEST_CANCELED ||
+        (queryResult.code == TRI_ERROR_QUERY_KILLED)) {
+      return Result(TRI_ERROR_REQUEST_CANCELED);
+    }
+    return Result(queryResult.code,
+                  std::string("Error fetching user defined AQL functions with this query: ") +
+                  queryResult.details);
+  }
+
+  auto usersFunctionsSlice = queryResult.result->slice();
+
+  if (!usersFunctionsSlice.isArray()) {
+    return Result(TRI_ERROR_INTERNAL, "bad query result for AQL User Functions");
+  }
+
+  result.openArray();
+  std::string tmp;
+  for (auto const& it : VPackArrayIterator(usersFunctionsSlice)) {
+    VPackSlice resolved;
+    if (it.isExternal()) {
```

Comment:
no need to check if the Slice is of type External, because resolveExternal will do this already.